# CM4

In [38]:
import pandas as pd
import numpy as np
from matplotlib import pyplot as plt
import seaborn as sns
from sklearn.neighbors import KNeighborsClassifier
from sklearn import metrics
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import train_test_split
from scipy import stats

## Iris

In [22]:
df_iris = pd.read_csv('iris_dataset_missing.csv')

In [23]:
df_iris['sepal_width'].interpolate(method='linear', direction = 'forward', inplace=True) 
df_iris['petal_length'].interpolate(method='linear', direction = 'forward', inplace=True) 
# https://www.w3resource.com/python-exercises/pandas/missing-values/python-pandas-missing-values-exercise-15.php

In [50]:
def minmax_scaling(a):
    xx=a
    a_std = (xx- xx.min(axis = 0)) / (xx.max(axis = 0) - xx.min(axis = 0))
    return a_std

In [25]:
def zscore(a):
    xx=a
    a_std = (xx- np.mean(xx)) / (np.std(xx))
    return a_std

In [52]:
mf_iris=df_iris.copy()
mf_iris['sepal_length']=minmax_scaling(mf_iris['sepal_length'])
mf_iris['sepal_width']=minmax_scaling(mf_iris['sepal_width'])
mf_iris['petal_length']=minmax_scaling(mf_iris['petal_length'])
mf_iris['petal_width']=minmax_scaling(mf_iris['petal_width'])

In [53]:
zf_iris=df_iris.copy()
zf_iris['sepal_length']=zscore(zf_iris['sepal_length'])
zf_iris['sepal_width']=zscore(zf_iris['sepal_width'])
zf_iris['petal_length']=zscore(zf_iris['petal_length'])
zf_iris['petal_width']=zscore(zf_iris['petal_width'])

### Without normalization 

In [54]:
train, test = train_test_split(df_iris, test_size = 0.2, stratify = df_iris['species'], random_state = 98)
X_train = train[['sepal_length','sepal_width','petal_length','petal_width']]
y_train = train.species
X_test = test[['sepal_length','sepal_width','petal_length','petal_width']]
y_test = test.species
counter = 0
scoresMean = np.zeros(35)
scoresStd  = np.zeros(35)
for k in range(1,36):
    classifier = KNeighborsClassifier(n_neighbors = k)
    classifier = classifier.fit(X_train,y_train)
    scores = cross_val_score(classifier, X_train, y_train, cv=5)
    scoresMean[counter] = scores.mean()
    scoresStd[counter] = scores.std()
    counter = counter + 1
    print('K=',k,'and mean of accuracy is',scores.mean(),'and standard deviation is',scores.std(),'.')

K= 1 and mean of accuracy is 0.9058823529411765 and standard deviation is 0.059988464865797474 .
K= 2 and mean of accuracy is 0.9294117647058823 and standard deviation is 0.04401949866792873 .
K= 3 and mean of accuracy is 0.9169117647058822 and standard deviation is 0.05971747207131458 .
K= 4 and mean of accuracy is 0.9169117647058822 and standard deviation is 0.05971747207131458 .
K= 5 and mean of accuracy is 0.9051470588235293 and standard deviation is 0.07023502315446797 .
K= 6 and mean of accuracy is 0.9051470588235293 and standard deviation is 0.07947981833453918 .
K= 7 and mean of accuracy is 0.9051470588235293 and standard deviation is 0.09531635542999299 .
K= 8 and mean of accuracy is 0.8816176470588235 and standard deviation is 0.07383758947431292 .
K= 9 and mean of accuracy is 0.9169117647058822 and standard deviation is 0.07958858310432242 .
K= 10 and mean of accuracy is 0.9404411764705882 and standard deviation is 0.05263391225933319 .
K= 11 and mean of accuracy is 0.916911

### By minmax

In [55]:
train, test = train_test_split(mf_iris, test_size = 0.2, stratify = mf_iris['species'], random_state = 98)
#corrmat = train.corr()
#sns.heatmap(corrmat, annot = True, square = True);
X_train = train[['sepal_length','sepal_width','petal_length','petal_width']]
y_train = train.species
X_test = test[['sepal_length','sepal_width','petal_length','petal_width']]
y_test = test.species
counter = 0
scoresMean = np.zeros(35)
scoresStd  = np.zeros(35)
for k in range(1,36):
    classifier = KNeighborsClassifier(n_neighbors = k)
    classifier = classifier.fit(X_train,y_train)
    scores = cross_val_score(classifier, X_train, y_train, cv=5)
    scoresMean[counter] = scores.mean()
    scoresStd[counter] = scores.std()
    counter = counter + 1
    print('K=',k,'and mean of accuracy is',scores.mean(),'and standard deviation is',scores.std(),'.')

K= 1 and mean of accuracy is 0.8941176470588236 and standard deviation is 0.06859943405700356 .
K= 2 and mean of accuracy is 0.9058823529411765 and standard deviation is 0.07979211744853258 .
K= 3 and mean of accuracy is 0.8933823529411764 and standard deviation is 0.06810903633476126 .
K= 4 and mean of accuracy is 0.9169117647058822 and standard deviation is 0.028249077518086103 .
K= 5 and mean of accuracy is 0.9169117647058822 and standard deviation is 0.05971747207131458 .
K= 6 and mean of accuracy is 0.9286764705882351 and standard deviation is 0.04384721034081052 .
K= 7 and mean of accuracy is 0.9404411764705882 and standard deviation is 0.05263391225933319 .
K= 8 and mean of accuracy is 0.9286764705882351 and standard deviation is 0.04384721034081052 .
K= 9 and mean of accuracy is 0.9286764705882351 and standard deviation is 0.04384721034081052 .
K= 10 and mean of accuracy is 0.9051470588235293 and standard deviation is 0.059572438530967935 .
K= 11 and mean of accuracy is 0.91691

### By zscore

In [56]:
from sklearn.model_selection import train_test_split
import seaborn as sns
from sklearn.neighbors import KNeighborsClassifier
from sklearn import metrics
from sklearn.model_selection import cross_val_score
from sklearn.preprocessing import StandardScaler

train, test = train_test_split(zf_iris, test_size = 0.2, stratify = zf_iris['species'], random_state = 98)
#corrmat = train.corr()
#sns.heatmap(corrmat, annot = True, square = True);
X_train = train[['sepal_length','sepal_width','petal_length','petal_width']]
y_train = train.species
X_test = test[['sepal_length','sepal_width','petal_length','petal_width']]
y_test = test.species
counter = 0
scoresMean = np.zeros(35)
scoresStd  = np.zeros(35)
for k in range(1,36):
    classifier = KNeighborsClassifier(n_neighbors = k)
    classifier = classifier.fit(X_train,y_train)
    scores = cross_val_score(classifier, X_train, y_train, cv=5)
    scoresMean[counter] = scores.mean()
    scoresStd[counter] = scores.std()
    counter = counter + 1
    print('K=',k,'and mean of accuracy is',scores.mean(),'and standard deviation is',scores.std(),'.')

K= 1 and mean of accuracy is 0.8941176470588236 and standard deviation is 0.06859943405700356 .
K= 2 and mean of accuracy is 0.9058823529411765 and standard deviation is 0.07979211744853258 .
K= 3 and mean of accuracy is 0.8823529411764707 and standard deviation is 0.08318903308077032 .
K= 4 and mean of accuracy is 0.9176470588235294 and standard deviation is 0.059988464865797474 .
K= 5 and mean of accuracy is 0.9294117647058823 and standard deviation is 0.06859943405700356 .
K= 6 and mean of accuracy is 0.9411764705882352 and standard deviation is 0.05261336417646565 .
K= 7 and mean of accuracy is 0.9294117647058823 and standard deviation is 0.06859943405700356 .
K= 8 and mean of accuracy is 0.9176470588235294 and standard deviation is 0.059988464865797474 .
K= 9 and mean of accuracy is 0.9058823529411765 and standard deviation is 0.07979211744853258 .
K= 10 and mean of accuracy is 0.9058823529411765 and standard deviation is 0.059988464865797474 .
K= 11 and mean of accuracy is 0.9294

## Heart disease

In [57]:
df = pd.read_csv('heart_disease_missing.csv')
if(df.isnull().values.any()):
    df = df.fillna(method = 'ffill')
df = df[(np.abs(stats.zscore(df)) < 4).all(axis=1)]
df = df[[ 'cp', 'thalach', 'exang', 'oldpeak','','target']]
df['oldpeak'].interpolate(method='linear', direction = 'forward', inplace=True) 

### Without normalization 

In [58]:
X = df.iloc[:,:-1].values
y = df.iloc[:,4].values
X_t, X_test, y_t, y_test =  train_test_split(X,y,test_size = 0.2, random_state= 98)
X_train, X_val, y_train, y_val =  train_test_split(X_t,y_t,test_size = 0.1, random_state= 98)
for k in range(1,36):
    classifier = KNeighborsClassifier(n_neighbors = k)
    classifier = classifier.fit(X_train,y_train)
    y_pred1 = classifier.predict(X_val)
    accuracy = metrics.accuracy_score(y_val, y_pred1)
    print('K=',k,'and Validation accuracy: {:.2f}'.format(accuracy),'.')
# https://www.kdnuggets.com/2019/07/classifying-heart-disease-using-k-nearest-neighbors.html/2

K= 1 and Validation accuracy: 0.71 .
K= 2 and Validation accuracy: 0.65 .
K= 3 and Validation accuracy: 0.71 .
K= 4 and Validation accuracy: 0.71 .
K= 5 and Validation accuracy: 0.71 .
K= 6 and Validation accuracy: 0.71 .
K= 7 and Validation accuracy: 0.71 .
K= 8 and Validation accuracy: 0.71 .
K= 9 and Validation accuracy: 0.76 .
K= 10 and Validation accuracy: 0.76 .
K= 11 and Validation accuracy: 0.88 .
K= 12 and Validation accuracy: 0.82 .
K= 13 and Validation accuracy: 0.76 .
K= 14 and Validation accuracy: 0.76 .
K= 15 and Validation accuracy: 0.76 .
K= 16 and Validation accuracy: 0.76 .
K= 17 and Validation accuracy: 0.76 .
K= 18 and Validation accuracy: 0.71 .
K= 19 and Validation accuracy: 0.71 .
K= 20 and Validation accuracy: 0.71 .
K= 21 and Validation accuracy: 0.71 .
K= 22 and Validation accuracy: 0.71 .
K= 23 and Validation accuracy: 0.76 .
K= 24 and Validation accuracy: 0.76 .
K= 25 and Validation accuracy: 0.76 .
K= 26 and Validation accuracy: 0.76 .
K= 27 and Validation 

In [59]:
from sklearn.preprocessing import StandardScaler
X = df.iloc[:,:-1].values
y = df.iloc[:,4].values
X_t, X_test, y_t, y_test =  train_test_split(X,y,test_size = 0.2, random_state= 98)
X_train, X_val, y_train, y_val =  train_test_split(X_t,y_t,test_size = 0.1, random_state= 98)
X_train = zscore(X_train)
X_test = zscore(X_test)
X_val = zscore(X_val)
for k in range(1,36):
    classifier = KNeighborsClassifier(n_neighbors = k)
    classifier = classifier.fit(X_train,y_train)
    y_pred1 = classifier.predict(X_val)
    accuracy = metrics.accuracy_score(y_val, y_pred1)
    print('K=',k,'and Validation accuracy: {:.2f}'.format(accuracy),'.')

K= 1 and Validation accuracy: 0.59 .
K= 2 and Validation accuracy: 0.71 .
K= 3 and Validation accuracy: 0.59 .
K= 4 and Validation accuracy: 0.53 .
K= 5 and Validation accuracy: 0.65 .
K= 6 and Validation accuracy: 0.65 .
K= 7 and Validation accuracy: 0.65 .
K= 8 and Validation accuracy: 0.59 .
K= 9 and Validation accuracy: 0.71 .
K= 10 and Validation accuracy: 0.65 .
K= 11 and Validation accuracy: 0.71 .
K= 12 and Validation accuracy: 0.71 .
K= 13 and Validation accuracy: 0.71 .
K= 14 and Validation accuracy: 0.71 .
K= 15 and Validation accuracy: 0.65 .
K= 16 and Validation accuracy: 0.65 .
K= 17 and Validation accuracy: 0.65 .
K= 18 and Validation accuracy: 0.65 .
K= 19 and Validation accuracy: 0.65 .
K= 20 and Validation accuracy: 0.65 .
K= 21 and Validation accuracy: 0.65 .
K= 22 and Validation accuracy: 0.65 .
K= 23 and Validation accuracy: 0.65 .
K= 24 and Validation accuracy: 0.59 .
K= 25 and Validation accuracy: 0.65 .
K= 26 and Validation accuracy: 0.59 .
K= 27 and Validation 

In [60]:
from sklearn.preprocessing import StandardScaler
X = df.iloc[:,:-1].values
y = df.iloc[:,4].values
X_t, X_test, y_t, y_test =  train_test_split(X,y,test_size = 0.2, random_state= 98)
X_train, X_val, y_train, y_val =  train_test_split(X_t,y_t,test_size = 0.1, random_state= 98)
X_train = minmax_scaling(X_train)
X_test = minmax_scaling(X_test)
X_val = minmax_scaling(X_val)
for k in range(1,36):
    classifier = KNeighborsClassifier(n_neighbors = k)
    classifier = classifier.fit(X_train,y_train)
    y_pred1 = classifier.predict(X_val)
    accuracy = metrics.accuracy_score(y_val, y_pred1)
    print('K=',k,'and Validation accuracy: {:.2f}'.format(accuracy),'.')

K= 1 and Validation accuracy: 0.71 .
K= 2 and Validation accuracy: 0.59 .
K= 3 and Validation accuracy: 0.65 .
K= 4 and Validation accuracy: 0.65 .
K= 5 and Validation accuracy: 0.71 .
K= 6 and Validation accuracy: 0.71 .
K= 7 and Validation accuracy: 0.76 .
K= 8 and Validation accuracy: 0.71 .
K= 9 and Validation accuracy: 0.76 .
K= 10 and Validation accuracy: 0.76 .
K= 11 and Validation accuracy: 0.76 .
K= 12 and Validation accuracy: 0.76 .
K= 13 and Validation accuracy: 0.76 .
K= 14 and Validation accuracy: 0.76 .
K= 15 and Validation accuracy: 0.76 .
K= 16 and Validation accuracy: 0.76 .
K= 17 and Validation accuracy: 0.76 .
K= 18 and Validation accuracy: 0.76 .
K= 19 and Validation accuracy: 0.76 .
K= 20 and Validation accuracy: 0.82 .
K= 21 and Validation accuracy: 0.82 .
K= 22 and Validation accuracy: 0.82 .
K= 23 and Validation accuracy: 0.82 .
K= 24 and Validation accuracy: 0.76 .
K= 25 and Validation accuracy: 0.76 .
K= 26 and Validation accuracy: 0.76 .
K= 27 and Validation 

Based on the results, for Iris data base accuracy of 94 % is reached with and without normalization. But with normalization number of neighborhoods is lower than without normalization. But for heart disease data base for 4 features, without normalization, the accuracy is better. 

## Refrences
https://www.kdnuggets.com/2019/07/classifying-heart-disease-using-k-nearest-neighbors.html/2
https://towardsdatascience.com/exploring-classifiers-with-python-scikit-learn-iris-dataset-2bcb490d2e1b
https://scikit-learn.org/stable/modules/cross_validation.html#cross-validation